File: sdk/go/resolvers/registrationrequirements.go
package resolvers

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"

	"github.com/RAMP-Protocol/protocol/sdk/go/helpers"
)

// ErrExchangeNotPermitted signals that the deployment's Allow overlay excluded
// this domain. It is raised BEFORE anything is dialled, so it says nothing about
// whether the Exchange exists or answers — only that this deployment declined to
// ask. Distinct from a transport failure because the remedy is a configuration
// change here, not a retry.
var ErrExchangeNotPermitted = errors.New("resolvers: exchange domain not permitted by policy")

// ErrManifestNotExchange signals that the document served at the domain's
// well-known path describes some other role. Registration requirements are an
// Exchange's to publish, so a manifest claiming to be an agent, a broker or a
// publisher is refused rather than read for members it has no business carrying.
// A manifest naming no role at all is refused the same way: the field is required
// by the contract, and treating an absent role as "probably an Exchange" would
// make the check advisory.
var ErrManifestNotExchange = errors.New("resolvers: well-known manifest does not describe an exchange")

// RegistrationRequirements is what one Exchange asks of a registration: the terms
// revision that submitting one accepts, and the schema its registration_data must
// match.
//
// Both members are optional in the contract, and their absence is a normal answer
// rather than a failure — an Exchange that publishes neither accepts registration
// data uninspected and records no terms acceptance.
type RegistrationRequirements struct {
	// TermsDigest is the manifest's terms_digest, nil when the Exchange publishes
	// none. Copy it onto RegisterRequest.terms_digest unchanged: the request
	// signature covers the echo, and that echo is the durable record of which
	// terms revision the operator accepted.
	//
	// A pointer rather than a string because absent and empty are different
	// answers on the wire, and only the pointer can carry the difference.
	TermsDigest *string

	// Schema validates registration_data before anything is signed. It is nil in
	// TWO cases — the Exchange publishes none, and the Exchange publishes one this
	// SDK refuses — and Verdict is what tells them apart. Both are deliberately
	// the same VALUE, because a nil *helpers.RegistrationSchema reports no
	// failures and that is the behaviour the contract requires of a client in both
	// cases: a local check that cannot run must not become a local veto.
	Schema *helpers.RegistrationSchema

	// Verdict is the SDK's answer for the published schema. SchemaNotPublished is
	// the ordinary absent case; SchemaAccepted means Schema is usable; anything
	// else names why a published schema was refused, which is worth logging and is
	// never worth refusing the registration over.
	Verdict helpers.SchemaVerdict
}

// RegistrationRequirementsReader reads an Exchange's published registration
// requirements. It is an interface so the client package can accept one without
// depending on this constructor, and so a test can drive registration without
// standing up a manifest server.
type RegistrationRequirementsReader interface {
	ResolveRegistrationRequirements(ctx context.Context, exchange string) (RegistrationRequirements, error)
}

// WellKnownRequirementsReader reads registration requirements out of an
// Exchange's own /.well-known/ramp.json.
//
// # Every read is a fresh fetch, and that is the whole design
//
// The protocol requires it in as many words: a registering client MUST read the
// terms digest from a freshly fetched manifest rather than a cached copy. A
// cached ENDPOINT is fine — a wrong one fails loudly — but a cached DIGEST is
// not, because a client cannot detect staleness locally, so a warm cache would
// make it echo a value the Exchange has already stopped accepting and retry the
// same refusal until the cache expired.
//
// That is why this is a SEPARATE reader rather than a third face on
// WellKnownEndpointResolver. That resolver is built out of exactly the mechanism
// this value may not touch — a per-host LRU with a TTL, and single-flight
// coalescing on top of it — and it exposes no bypass. Adding a member to the
// document it decodes would satisfy the rule's letter at the one point it fails
// in practice: the first caller to reuse the SDK's own manifest path would be
// handed a digest minutes old with nothing in the API to warn them. A reader that
// holds no document cache leaves no cache slot to reuse, which is the same
// structural argument the report leg makes about configuration — the rule holds
// because there is nowhere for the wrong answer to come from, not because callers
// remember.
//
// Reading the schema out of the same fresh bytes then costs nothing extra and
// removes a failure mode of its own: a stale local schema cannot refuse a payload
// the Exchange would have taken, because there is no stale local schema.
//
// # What is deliberately NOT held
//
// Not even the compiled validator. Compiling an attacker-authored schema is
// bounded but not free, and an application registering repeatedly at the same
// Exchange may reasonably want to pay that once — but memoising it is a cache,
// which this tier does not do by default, and the useful key for it (the digest
// of the served bytes) is a property of a deployment's threat model rather than
// of the protocol. An application that wants it wraps this reader.
type WellKnownRequirementsReader struct {
	http   *http.Client
	scheme string
	allow  func(domain string) bool
}

// NewWellKnownRequirementsReader returns a reader over WellKnownOptions. Zero-value
// options are safe defaults: https scheme and the SSRF-GUARDED client.
//
// The domain is CALLER-NAMED — an agent registers at whichever Exchange it means
// to transact with, and that domain routinely arrives at runtime rather than from
// configuration — so this is the same threat shape as the endpoint resolver's
// fetch and takes the same guarded default, NOT http.DefaultClient. A deployment
// that must reach a private or loopback Exchange injects its own client via
// opts.HTTP or opts out through the SKIP_SSRF / ALLOW_INSECURE env flags.
//
// opts.TTL and opts.Now are accepted and ignored: this reader caches nothing, so
// it has no freshness to compute. They are not errors, so one WellKnownOptions
// value can build every face in this package.
func NewWellKnownRequirementsReader(opts WellKnownOptions) *WellKnownRequirementsReader {
	client := opts.HTTP
	if client == nil {
		client = NewGuardedClientFromEnv()
	}
	scheme := opts.Scheme
	if scheme == "" {
		scheme = "https"
	}
	return &WellKnownRequirementsReader{http: client, scheme: scheme, allow: opts.Allow}
}

// ResolveRegistrationRequirements fetches exchange's manifest and reports what it
// asks of a registration. The answer is never served from a cache; see the type
// doc for why that is the point rather than an omission.
//
// Two refusals come before anything is dialled, and both are here rather than at
// the call site for the reason the endpoint resolver gives for its own: they are
// properties of building this URL and of this deployment's policy, not of any one
// caller's plans, and a second caller — or one that forgot — would otherwise fetch
// whatever domain an authenticated agent named.
//
// The SHAPE predicate is IsBareDomain, the contract's rule, and not the routing
// tier's IsBareHost. The two are kept apart deliberately and this leg is where the
// difference bites: nothing upstream of here has run the contract's rule, and the
// URL below is built by concatenation, so a value carrying a path or userinfo
// would choose WHAT is fetched rather than merely where from. A trailing root dot,
// a leading or trailing hyphen, an underscore and a bracketed IPv6 literal are all
// usable hosts that the wire rule refuses.
//
// A refused schema is never an error. The verdict is returned alongside a nil
// Schema, because the contract requires a client that cannot check locally to send
// anyway and let the Exchange decide.
func (r *WellKnownRequirementsReader) ResolveRegistrationRequirements(
	ctx context.Context, exchange string,
) (RegistrationRequirements, error) {
	if !helpers.IsBareDomain(exchange) {
		return RegistrationRequirements{}, fmt.Errorf(
			"resolvers: registration requirements: %w: %q is not a bare domain",
			helpers.ErrInvalidHost, exchange)
	}
	if r.allow != nil && !r.allow(exchange) {
		return RegistrationRequirements{}, fmt.Errorf("%w: %q", ErrExchangeNotPermitted, exchange)
	}
	// Bounded whatever client was injected: WellKnownOptions.HTTP admits one with
	// no timeout at all, and a caller may hold no deadline of its own. The ceiling
	// is the endpoint resolver's, so both manifest fetches wait the same.
	fetchCtx, cancel := context.WithTimeout(ctx, maxManifestFetch)
	defer cancel()
	url := r.scheme + "://" + exchange + "/.well-known/ramp.json"
	doc, err := fetchWellKnownDoc(fetchCtx, r.http, url)
	if err != nil {
		return RegistrationRequirements{}, fmt.Errorf(
			"resolvers: registration requirements for %q: %w", exchange, err)
	}
	if !doc.describesExchange() {
		return RegistrationRequirements{}, fmt.Errorf("%w: host=%q", ErrManifestNotExchange, exchange)
	}
	reqs := RegistrationRequirements{TermsDigest: doc.TermsDigest, Verdict: helpers.SchemaNotPublished}
	// The schema is compiled from the bytes AS SERVED, because every cap the rules
	// state is defined over those bytes. json.RawMessage is the sub-document the
	// decoder saw, so nothing has re-encoded it on the way here.
	if raw := doc.registrationSchemaBytes(); len(raw) > 0 {
		reqs.Schema, reqs.Verdict = helpers.CompileRegistrationSchema(raw)
	}
	return reqs, nil
}

// roleExchange is the proto-JSON rendering of Role.ROLE_EXCHANGE. A manifest may
// also carry the enum's number, which proto-JSON permits, so both are accepted.
const roleExchange = "ROLE_EXCHANGE"

// roleExchangeNumber is Role.ROLE_EXCHANGE's field number, accepted because
// proto-JSON allows an enum to travel as its number.
const roleExchangeNumber = "2"

// describesExchange reports whether the manifest names the Exchange role. An
// absent or unparsable role is not an Exchange: the contract makes the field
// required, and reading silence as assent would leave the check advisory.
func (d *wellKnownDoc) describesExchange() bool {
	var s string
	if err := json.Unmarshal(d.Role, &s); err == nil {
		return s == roleExchange
	}
	var n json.Number
	if err := json.Unmarshal(d.Role, &n); err == nil {
		return n.String() == roleExchangeNumber
	}
	return false
}

// registrationSchemaBytes returns the published data_schema exactly as served, or
// nil when the block or the member is absent. Absent and blank are the same answer
// to the caller — CompileRegistrationSchema reads either as "publishes none" — but
// the two are distinguished here anyway so a blank member reaches the rule that
// defines blankness rather than a rule invented at this layer.
func (d *wellKnownDoc) registrationSchemaBytes() []byte {
	if d.AccountRegistration == nil {
		return nil
	}
	return d.AccountRegistration.DataSchema
}

File: sdk/ts/resolvers/registration-requirements.ts
// Reading what one Exchange asks of a registration — the terms revision that
// submitting one accepts, and the schema its registration_data must match — out of
// that Exchange's own /.well-known/ramp.json. TS port of the Go oracle
// (resolvers/registrationrequirements.go).
//
// EVERY READ IS A FRESH FETCH, and that is the whole design. The protocol requires
// it in as many words: a registering client MUST read the terms digest from a
// freshly fetched manifest rather than a cached copy. A cached ENDPOINT is fine — a
// wrong one fails loudly — but a cached DIGEST is not, because a client cannot
// detect staleness locally, so a warm cache would make it echo a value the Exchange
// has already stopped accepting and retry the same refusal until the cache expired.
//
// That is why this is a SEPARATE face rather than a third method on the endpoint
// resolver. That resolver is built out of exactly the mechanism this value may not
// touch — a per-host TTL cache with single-flight coalescing on top — and it exposes
// no bypass. A face that holds no document cache leaves no cache slot to reuse,
// which is what makes the rule structural rather than a convention callers remember.
//
// Not even the compiled validator is held. Memoising it is a cache, and the useful
// key for one is a property of a deployment's threat model rather than of the
// protocol; an application that wants it wraps this face.

import { isBareDomain } from "../src/hosts.ts";
import { invalidHost } from "../src/host-ref.ts";
import {
  compileRegistrationSchema,
  type RegistrationSchema,
  type SchemaVerdict,
} from "../src/regschema.ts";
import { DirectoryUnavailable, ExchangeNotPermitted, ManifestNotExchange } from "./errors.ts";
import { type FetchLike, defaultFetch, fetchStrict } from "./http.ts";

/** What one Exchange asks of a registration. Both members are optional in the
 * contract, and their absence is a normal answer rather than a failure. */
export interface RegistrationRequirements {
  /** The manifest's `terms_digest`, `undefined` when the Exchange publishes none.
   * Copy it onto `RegisterRequest.terms_digest` unchanged: the request signature
   * covers the echo, and that echo is the durable record of which terms revision
   * the operator accepted. */
  termsDigest: string | undefined;
  /** Validates `registration_data` before anything is signed. `null` in TWO cases
   * — the Exchange publishes none, and it publishes one this SDK refuses — and
   * `verdict` is what tells them apart. Both are deliberately the same VALUE,
   * because the contract requires a client that cannot check locally to send
   * anyway: a local check that cannot run must not become a local veto. */
  schema: RegistrationSchema | null;
  /** The SDK's answer for the published schema. `not_published` is the ordinary
   * absent case; `accepted` means `schema` is usable; anything else names why a
   * published schema was refused, which is worth logging and is never worth
   * refusing the registration over. */
  verdict: SchemaVerdict;
}

/** The well-known registration-requirements face, named for the concrete reader
 * the oracle exposes (`resolvers.WellKnownRequirementsReader`). */
export interface WellKnownRequirementsReader {
  resolveRegistrationRequirements(exchange: string): Promise<RegistrationRequirements>;
}

/** Options for the reader. `ttlMs` and `now` are deliberately absent: it caches
 * nothing, so it has no freshness to compute. */
export interface WellKnownRequirementsOptions {
  fetch?: FetchLike;
  /** Trust allowlist consulted BEFORE the fetch. A domain it rejects never
   * reaches the network. */
  allow?: (domain: string) => boolean;
  /** The URL scheme used to build `{scheme}://{domain}/.well-known/ramp.json`
   * (tests inject "http"). */
  scheme?: string;
}

/** Read registration requirements from an Exchange's own well-known manifest. */
export function createWellKnownRequirementsReader(
  opts: WellKnownRequirementsOptions = {},
): WellKnownRequirementsReader {
  const fetchFn: FetchLike = opts.fetch ?? defaultFetch;
  const scheme = opts.scheme && opts.scheme !== "" ? opts.scheme : "https";
  const allow = opts.allow;

  return {
    async resolveRegistrationRequirements(exchange: string): Promise<RegistrationRequirements> {
      // The SHAPE predicate is the contract's isBareDomain, not the routing tier's
      // isBareHost. The two are kept apart deliberately and this leg is where the
      // difference bites: nothing upstream has run the contract's rule, and the URL
      // below is built by concatenation, so a value carrying a path or userinfo
      // would choose WHAT is fetched rather than merely where from.
      if (!isBareDomain(exchange)) {
        throw invalidHost(exchange, "not a bare domain");
      }
      if (allow && !allow(exchange)) {
        throw new ExchangeNotPermitted(`exchange ${exchange} not permitted by policy`);
      }
      const url = `${scheme}://${exchange}/.well-known/ramp.json`;
      const body = await fetchStrict(fetchFn, url);
      let doc: unknown;
      try {
        doc = JSON.parse(body);
      } catch (err) {
        throw new DirectoryUnavailable(`manifest decode ${url}`, { cause: err });
      }
      if (doc === null || typeof doc !== "object" || Array.isArray(doc)) {
        throw new DirectoryUnavailable(`manifest at ${url} is not a JSON object`);
      }
      if (!describesExchange(doc)) {
        throw new ManifestNotExchange(`host=${exchange}`);
      }
      const digest = (doc as { terms_digest?: unknown }).terms_digest;
      const out: RegistrationRequirements = {
        termsDigest: typeof digest === "string" ? digest : undefined,
        schema: null,
        verdict: "not_published",
      };
      // Compiled from the bytes AS SERVED. Every cap the registration-schema rules
      // state is defined over those bytes, so the member is sliced out of the body
      // rather than re-serialised from the parsed value: JSON.parse discards
      // whitespace and reformats numbers, and a document that minifies under the
      // cap while exceeding it on the wire would be refused by an Exchange and
      // accepted here — the two-privately-chosen-limits failure the contract's
      // numbers exist to prevent.
      const raw = rawMember(body, "account_registration", "data_schema");
      if (raw !== undefined) {
        const compiled = compileRegistrationSchema(raw);
        out.schema = compiled.schema;
        out.verdict = compiled.verdict;
      }
      return out;
    },
  };
}

/** proto-JSON renders an enum as its name or its number, so both are accepted. An
 * absent or unrecognised role is not an Exchange. */
function describesExchange(doc: unknown): boolean {
  const role = (doc as { role?: unknown }).role;
  return role === "ROLE_EXCHANGE" || role === 2;
}

/** rawMember returns the EXACT served text of a nested object member, or
 * `undefined` when any step of the path is absent.
 *
 * It exists because JavaScript has no equivalent of a raw-message slice: the only
 * way to hand a nested member to a rule defined over the bytes AS SERVED is to
 * find its extent in the body. `body` has already parsed as JSON when this runs,
 * so the scan below is over known-well-formed input and needs no error path — it
 * returns `undefined` for anything it cannot walk, and the caller reads that as
 * "no schema published", which is the safe direction.
 */
function rawMember(body: string, ...path: string[]): string | undefined {
  let start = 0;
  for (let step = 0; step < path.length; step++) {
    const found = memberExtent(body, start, path[step] as string);
    if (found === undefined) return undefined;
    if (step === path.length - 1) return body.slice(found[0], found[1]);
    start = found[0];
  }
  return undefined;
}

/** memberExtent finds `key` in the object beginning at `from` and returns the
 * [start, end) of its VALUE. */
function memberExtent(body: string, from: number, key: string): [number, number] | undefined {
  let i = skipWhitespace(body, from);
  if (body[i] !== "{") return undefined;
  i = skipWhitespace(body, i + 1);
  if (body[i] === "}") return undefined;
  for (;;) {
    if (body[i] !== '"') return undefined;
    const keyEnd = skipString(body, i);
    const name = JSON.parse(body.slice(i, keyEnd)) as string;
    i = skipWhitespace(body, keyEnd);
    if (body[i] !== ":") return undefined;
    const valueStart = skipWhitespace(body, i + 1);
    const valueEnd = skipValue(body, valueStart);
    if (valueEnd === undefined) return undefined;
    if (name === key) return [valueStart, valueEnd];
    i = skipWhitespace(body, valueEnd);
    if (body[i] === ",") {
      i = skipWhitespace(body, i + 1);
      continue;
    }
    return undefined;
  }
}

/** RFC 8259 whitespace, and nothing wider — the same four bytes the blank-schema
 * rule counts. */
function skipWhitespace(body: string, i: number): number {
  while (i < body.length) {
    const c = body[i];
    if (c !== " " && c !== "\t" && c !== "\r" && c !== "\n") return i;
    i++;
  }
  return i;
}

/** skipString returns the index just past a JSON string starting at `i`. */
function skipString(body: string, i: number): number {
  i++; // the opening quote
  while (i < body.length) {
    const c = body[i];
    if (c === "\\") {
      i += 2;
      continue;
    }
    if (c === '"') return i + 1;
    i++;
  }
  return i;
}

/** skipValue returns the index just past the JSON value starting at `i`, or
 * `undefined` when the value does not terminate. */
function skipValue(body: string, i: number): number | undefined {
  const c = body[i];
  if (c === '"') return skipString(body, i);
  if (c === "{" || c === "[") {
    const close = c === "{" ? "}" : "]";
    let depth = 0;
    while (i < body.length) {
      const ch = body[i];
      if (ch === '"') {
        i = skipString(body, i);
        continue;
      }
      if (ch === "{" || ch === "[") depth++;
      else if (ch === "}" || ch === "]") {
        depth--;
        if (depth === 0) {
          return body[i] === close ? i + 1 : undefined;
        }
      }
      i++;
    }
    return undefined;
  }
  // A number, or one of the three literals: everything up to the next structural
  // character. The body has already parsed, so whatever lies here is well formed.
  const end = /[\s,}\]]/.exec(body.slice(i));
  return end === null ? body.length : i + end.index;
}
